### PR TITLE
save_variables: extras module to save variables to disk

### PR DESCRIPTION
**Pull request by dushyantahuja**
module:SAVE_VARIABLES, extras module to save variables to disk

Saves the arbitrary variables to disk so that it can be used across restarts. All stored variables are loaded into the printer.save_variables dict at startup and can be used in gcode macros.

Signed Off: Dushyant Ahuja <dusht.ahuja@gmail.com>

### CODE REVIEW

**Comment by dushyantahuja** [comment]
Not sure why I'm getting a build failed - as the log is not showing up for me - will check again tomorrow.

**Comment by dushyantahuja** [comment]
Whew - that was more difficult than it should have been !!

File: klippy/extras/save_variables.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
+# Get and Save arbitrary variables in the config file so that values can be kept
+# across restarts.
+#
+# This file may be distributed under the terms of the GNU GPLv3 license.
```
Comment (KevinOConnor) [inline klippy/extras/save_variables.py:4]: Please shrink the one line description to just one line.

There needs to be a copyright statement present - see the other Klipper files for an example layout.

File: klippy/extras/save_variables.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+import sys,os,logging, ConfigParser
+
+error = ConfigParser.Error
```
Comment (KevinOConnor) [inline klippy/extras/save_variables.py:8]: It's not a good idea to define global variables like this.  I recommend just using ConfigParser.Error explicitly in the locations where it is needed.

File: klippy/extras/save_variables.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+class SaveVariables:
+    error = ConfigParser.Error
```
Comment (KevinOConnor) [inline klippy/extras/save_variables.py:11]: Similar.

File: klippy/extras/save_variables.py [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    def __init__(self, config):
+        self.printer = config.get_printer()
+        self.gcode = self.printer.lookup_object('gcode')
+        self.name = config.get_name()
+        self.cfilename = config.get('filename')
+        self.filename = os.path.join(os.path.expanduser("~"), self.cfilename)
```
Comment (KevinOConnor) [inline klippy/extras/save_variables.py:17]: This path expansion doesn't look right.  If the user specifies an absolute path to a file then it should use that.  I think you want to use `os.path.expanduser(self.cfilename)`.

File: klippy/extras/save_variables.py [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+        self.allVariables = {}
+        self.loadVariables()
+        logging.info(str(self.allVariables))
```
Comment (KevinOConnor) [inline klippy/extras/save_variables.py:20]: When logging, it's important to provide context to the log message - so, for example, `logging.info("save_variable state: %s", self.allVariables)`.

File: klippy/extras/save_variables.py [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+    def loadVariables(self):
+        self.allVariables = {}
+        try:
+            self.variablefile = ConfigParser.ConfigParser()
+            self.variablefile.read(self.filename)
+            self.allVariables = dict(self.variablefile.items('Variables'))
+        except error as e:
+            msg = e.message + "\nUnable to parse existing variable file"
+            logging.exception(msg)
+            raise self.gcode.error(msg)
```
Comment (KevinOConnor) [inline klippy/extras/save_variables.py:34]: It's not necessary to include the error as part of the message report as the `logging.exception()` call already does that.  So, this should look something like `logging.exception("Unable to parse save_variables file")`.

File: klippy/extras/save_variables.py [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+    def cmd_SAVE_VARIABLE(self, params):
+        self.variable_name = self.gcode.get_str('VARIABLE', params)
+        self.variable_value = self.gcode.get_str('VALUE', params)
+        try:
+            self.variablefile = ConfigParser.ConfigParser()
+            self.variablefile.read(self.filename)
+        except error as e:
+            msg = "Unable to parse existing variable file"
+            logging.exception(msg)
+            raise self.gcode.error(msg)
```
Comment (KevinOConnor) [inline klippy/extras/save_variables.py:44]: Could this be replaced with a call to `self.loadVariables()`?

File: klippy/extras/save_variables.py [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+        except error as e:
+            msg = e.message + "\nUnable to save variable"
+            logging.exception(msg)
+            raise self.gcode.error(msg)
+        self.gcode.respond_info("Variable Saved")
+        self.loadVariables()
+        logging.info(str(self.allVariables))
+    def get_status(self, eventtime):
+        #return {'temp1':self.filename}
+        #logging.info("allVariables='%s'", repr(self.allVariables))
```
Comment (KevinOConnor) [inline klippy/extras/save_variables.py:58]: Please don't commit commented out code.

File: klippy/extras/save_variables.py [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+        return self.allVariables
```
Comment (KevinOConnor) [inline klippy/extras/save_variables.py:59]: I'd recommend doing something like `return {"variables": self.allVariables}` as then it becomes possible to add additional exports in the future without fear of conflicts.

File: docs/G-Codes.md [hunk 10]
```diff
@@ -579,3 +579,30 @@ been enabled:
...
+
+[gcode_macro T0]
+gcode:
+ACTIVATE_EXTRUDER extruder=extruder
+SAVE_VARIABLE VARIABLE=currentextruder VALUE=extruder
+
+[gcode_macro START_GCODE]
+gcode:
+ACTIVATE_EXTRUDER extruder={printer.save_variables.currentextruder}
+```
```
Comment (KevinOConnor) [inline docs/G-Codes.md:608]: I'd recommend against providing config file examples in docs/G-Codes.md as it is "off topic" for that documentation.  I'd add a blurb to docs/Command_Templates.md to describe the "printer variable" itself.  Also, I'd mention here that variables are always stored as Python strings.

File: klippy/extras/save_variables.py (hunk 1 above)
Comment (dushyantahuja) [reply klippy/extras/save_variables.py:4]: Done

File: klippy/extras/save_variables.py (hunk 2 above)
Comment (dushyantahuja) [reply klippy/extras/save_variables.py:8]: I had used configfile.py as a base for my code. Have removed this now.

File: klippy/extras/save_variables.py (hunk 4 above)
Comment (dushyantahuja) [reply klippy/extras/save_variables.py:17]: OK - the original intent was that the user provides only the filename; but this makes more sense.

File: klippy/extras/save_variables.py (hunk 5 above)
Comment (dushyantahuja) [reply klippy/extras/save_variables.py:20]: Done

File: klippy/extras/save_variables.py (hunk 7 above)
Comment (dushyantahuja) [reply klippy/extras/save_variables.py:44]: you're right - didn't realise that

File: klippy/extras/save_variables.py (hunk 8 above)
Comment (dushyantahuja) [reply klippy/extras/save_variables.py:58]: OK

File: klippy/extras/save_variables.py (hunk 9 above)
Comment (dushyantahuja) [reply klippy/extras/save_variables.py:59]: How would this work in terms of the printer variable - would `printer.save_variables.currentextruder` become `printer.save_variables.variables.currentextruder` ? I'd prefer not to do that in that case. I doubt I'd add more feature to this module in any case - it's simple and works. Which I think is great.

File: docs/G-Codes.md (hunk 10 above)
Comment (dushyantahuja) [reply docs/G-Codes.md]: OK - I just thought that putting an example here would make it easier for users to understand how this works. Will add something to Command_Templates.

**Comment by dushyantahuja** [comment]
Thanks for the detailed comments, Kevin.

File: klippy/extras/save_variables.py (hunk 6 above)
Comment (dushyantahuja) [reply klippy/extras/save_variables.py:32]: Done

**Comment by KevinOConnor** [comment]
Thanks. At a high-level is looks fine to me.  Some comments:

I'd prefer `get_status()` to not directly return the dictionary - instead use something like: `return {'variables': dict(self.allVariables)}`.  I know this can be a minor pain, but I'd rather be a little verbose now then have to deal with backwards incompatible breakage in the future.  For macros, I'd recommend adding something like `{% set svv = printer.save_variables.variables %}` to the top of the macro to avoid overly long lines.  Note that the get_status() method should always copy (eg, via `dict()`) the local dictionary as we don't want the local storage to get modified by an errant macro.

The `if os.path.isfile(self.filename):` checks do not look correct, as there are several errors that could occur besides file not present (eg, no permissions).  Instead, the `__init__()` code should use `try: ... except self.printer.command_error, e: raise config.error(str(e))` to make sure the error is propagating correctly.

The cmd_SAVE_VARIABLE code does not appear to be closing the opened file - that will lead to a file descriptor leak.  Also, the same function is assigning self.variable_name / self.variable_value where it should instead be using local variables (eg, variable_name / variable_value).

If you want to store the values as integer/float/string, then you could try saving them with `repr(variable_value)` and parse it out with `ast.literal_eval()`.

-Kevin

File: klippy/extras/save_variables.py (hunk 9 above)
Comment (dushyantahuja) [reply klippy/extras/save_variables.py:53]: I've made the change.

**Comment by dushyantahuja** [comment]
Thanks Kevin, I've made the changes recommended. The variables are now being converted to floats if possible.

**Comment by KevinOConnor** [comment]
Thanks.  The error handling doesn't look correct to me - checking for file existence (or similar) prior to a read or open does not ensure that the subsequent read/open will not fail.  Instead, it's necessary to handle an error on every os call itself.

Also, I recommend using `ast.literal_eval()` and `repr()` instead of trying to parse all values as floats.

How about something like this (totally untested):
```python
# Save arbitrary variables so that values can be kept across restarts.
#
# Copyright (C) 2020 Dushyant Ahuja <dusht.ahuja@gmail.com>
# Copyright (C) 2016-2020  Kevin O'Connor <kevin@koconnor.net>
#
# This file may be distributed under the terms of the GNU GPLv3 license.
import os, logging, ast, ConfigParser

class SaveVariables:
    def __init__(self, config):
        self.printer = config.get_printer()
        self.filename = os.path.expanduser(config.get('filename'))
        self.allVariables = {}
        try:
            self.loadVariables()
        except self.printer.command_error as e:
            raise config.error(str(e))
        gcode = self.printer.lookup_object('gcode')
        gcode.register_command('SAVE_VARIABLE', self.cmd_SAVE_VARIABLE,
                               desc=self.cmd_SAVE_VARIABLE_help)
    def loadVariables(self):
        allvars = {}
        varfile = ConfigParser.ConfigParser()
        try:
            varfile.read(self.filename)
            for name, val in varfile.items('Variables'):
                allvars[name] = ast.literal_eval(val)
        except:
            msg = "Unable to parse existing variable file"
            logging.exception(msg)
            raise self.printer.command_error(msg)
        self.allVariables = allvars
    cmd_SAVE_VARIABLE_help = "Save arbitrary variables to disk"
    def cmd_SAVE_VARIABLE(self, gcmd):
        varname = gcmd.get('VARIABLE')
        value = gcmd.get('VALUE')
        try:
            value = ast.literal_eval(value)
        except ValueError as e:
            raise gcmd.error("Unable to parse '%s' as a literal" % (value,))
        newvars = dict(self.allVariables)
        newvars[varname] = value
        # Write file
        varfile = ConfigParser.ConfigParser()
        varfile.add_section('Variables')
        for name, val in sorted(newvars.items()):
            varfile.set('Variables', name, repr(val))
        try:
            f = open(self.filename, "w")
            varfile.write(f)
            f.close()
        except:
            msg = "Unable to save variable"
            logging.exception(msg)
            raise gcmd.error(msg)
        gcmd.respond_info("Variable Saved")
        self.loadVariables()
    def get_status(self, eventtime):
        return {'variables': dict(self.allVariables)}

def load_config(config):
    return SaveVariables(config)
```

**Comment by KevinOConnor** [comment]
Any further updates on this?

-Kevin

**Comment by dushyantahuja** [comment]
I'm sorry - am in the middle of moving countries - this has gone on the
back-burner, and I will probably not be able to get to it for at least the
next 4-6 weeks.

------------------
Dushyant Ahuja

**Comment by dushyantahuja** [comment]
Hi Kevin, I've made changes recommended; can you please review and advise now.

**Comment by KevinOConnor** [comment]
Thanks.  FYI, with the next release of Klipper coming up in a few days (see #3363) - this is something I'd look to do an in-depth review and merge after the release is made.  That's likely going to be in a couple of weeks.

Cheers,
-Kevin

**Comment by dushyantahuja** [comment]
Sounds good, let me know if anything else needs to be modified.

**Comment by KevinOConnor** [comment]
Thanks.  I got a chance to review the code.  Some high-level comments:
* The error handling doesn't look correct to me.  Any time `open()` is called, it's necessary to check for error results (which isn't done in the init method).  Also, I don't understand the use of `isfile()` - as it's basically never safe to use that call.  Is there a reason to check for file existence as opposed to just always reading the results?
* The documentation states that values are converted to floats if they can be - but is that the case?  It looks to me like the code would just result in an error if one tried to save a string.  FWIW, I don't think it's a good idea to try to convert to floats - I'd use `ast.literal_eval()` and `repr()` to make sure values are stored exactly as the user specifies them (strings, ints, or floats).

-Kevin

**Comment by dushyantahuja** [comment]
I'm sorry, I've reached the end of my python expertise. I'm afraid I don't know best practices, as a matter of fact, I don't know what I don't know. Am closing this pull request. If anyone else wants to take it up, they are more than welcome. 

Thanks for all the help.

**Comment by KevinOConnor** [comment]
Okay.  I made the remaining changes and pushed this as commit d7053f6e.  Thanks for working on this and I apologize if the process was frustrating.

-Kevin

**Comment by dushyantahuja** [comment]
Not at all @KevinOConnor - the issues were due to my shortcomings and ignorance of python. Thanks to your help I have learned a lot.  Do I need to do anything more for this to be merged?

**Comment by KevinOConnor** [comment]
It should be merged as of commit d7053f6.  If it's missing anything, let me know.

-Kevin

**Pull request closed without merging**